Add support for stream dependency and priority weight

mmopenet opened the pull request:
PR for #2685

vietj commented on src/main/java/io/vertx/core/http/HttpClientRequest.java:
```diff
@@ -387,4 +387,15 @@ default int streamId() {
+  /**
+   * Sets the priority weight of the associated stream
+   * @param weight The weight priority weight or this requet's stream  
+   */
```
make it fluent

vietj commented on src/main/java/io/vertx/core/http/HttpServerResponse.java:
```diff
@@ -472,4 +472,15 @@ default void reset() {
+   * Sets the priority weight of the associated stream
+   * @param weight The weight priority weight or this requet's stream  
+   */
+  void setWeight(short weight);
```
make it fluent

vietj commented on src/main/java/io/vertx/core/http/impl/Http2ClientConnection.java:
```diff
@@ -132,6 +133,14 @@ private void recycle() {
   }
 
+  @Override
+  public void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers, int streamDependency, short weight, boolean exclusive, int padding, boolean endOfStream) throws Http2Exception {
```
it should be synchronized

vietj commented:
I think we need tests that use directly the netty API, see Http2ServerTest and Http2ClientTest as well

vietj commented on src/main/java/io/vertx/core/http/HttpServerResponse.java:
```diff
@@ -472,4 +472,15 @@ default void reset() {
+   * Sets the dependecy of associated stream
+   * @param The identifier of the HTTP/2 stream this request's stream depends on
+   */
+  void setStreamDependency(int streamDependency);
```
rather `streamId` for argument

vietj commented:
I think we are missing two things

- make the API dynamic such that changing the priority or dependency actually sends a new priority frame
- ability to react on priority change, so have one the API an `priorityUpdateHandler` that is called when the priority of a stream has changed.

instead of having a two fields : identity and dependency, we would have instead a `StreamPriority` data object that is used it can be changed with a single call. The priority update handler then would be an `Handler<StreamPriority>`

mmopenet commented:
How about sending priority frame if the priority or dependency changes after HEADERS frame was sent?
If the priority or dependency is changed before HEADERS frame is sent then the priority and stream dependency would be set in HEADERS frame (either in request or response).

vietj commented:
@mmopenet you are totally right, before the stream is sent, it should be part of headers , after it should be a priority frame. The request/response should allow to set an handler to be aware of priority changes after the stream has started

mmopenet commented:
Hi @vietj, I think I have addressed all the comments.

vietj commented:
@mmopenet can you sign the Eclipse Contributor Agreement and the commits to have a valid PR ?

vietj commented on src/main/java/io/vertx/core/http/HttpClientRequest.java:
```diff
@@ -387,4 +387,21 @@ default int streamId() {
+   * @return The priority of this requet's stream  
+   */
+  default StreamPriority getStreamPriority() {
+      return StreamPriority.DEFAULT;
```
return `null` instead when there is no priority

vietj commented on src/main/java/io/vertx/core/http/HttpClientResponse.java:
```diff
@@ -153,4 +153,17 @@
   HttpClientRequest request();
+
+  /**
+   * @return The priority of associated HTTP/2 stream  
```
`otherwise {@code null}`

vietj commented on src/main/java/io/vertx/core/http/HttpServerRequest.java:
```diff
@@ -328,4 +329,18 @@ default HttpServerRequest bodyHandler(@Nullable Handler<Buffer> bodyHandler) {
+   * @return The priority of associated stream  
+   */
+  default StreamPriority getStreamPriority() {
+      return StreamPriority.DEFAULT;
```
return `null` by default

vietj commented on src/main/java/io/vertx/core/http/StreamPriority.java:
```diff
@@ -0,0 +1,122 @@
+@DataObject
+public class StreamPriority {
+    
+    public static final StreamPriority DEFAULT = new StreamPriority();
```
not needed

mmopenet replied:
This is still needed for default value in the VertxHttp2Stream.

vietj commented on src/main/java/io/vertx/core/http/StreamPriority.java:
```diff
@@ -0,0 +1,122 @@
+    }
+    
+    public StreamPriority(JsonObject json) {
+        this.weight = json.getInteger("weight", (int)Http2CodecUtil.DEFAULT_PRIORITY_WEIGHT).shortValue();
```
generated instead of converter (with code generation) so naming follows the converter convention.

mmopenet replied:
This one is a bit problematic. I intentionally have not added setters to StreamPriority, so you can't do something like:
req.getStreamPriority().setWeight(4)
req.getStreamPriority().setStreamDependecy(5)
If you do this the priority frame would not be sent which would be confusing.

If I do not have setters or setter access is not public the generator generates converter like this:
   static void fromJson(Iterable<java.util.Map.Entry<String, Object>> json, StreamPriority obj) {
    for (java.util.Map.Entry<String, Object> member : json) {
      switch (member.getKey()) {
        case "exclusive":
          break;
        case "streamDependency":
          break;
        case "weight":
          break;
      }
    }
  }

Therefore I think it would be better to not use the generator or update it so it works with package accessible methods.

vietj commented on src/main/java/io/vertx/core/http/StreamPriority.java:
```diff
@@ -0,0 +1,122 @@
+    }
+    
+    public JsonObject toJson() {
+      JsonObject json = new JsonObject();
```
use generated converter instead

vietj commented on src/main/java/io/vertx/core/http/impl/Http2ConnectionBase.java:
```diff
@@ -193,8 +194,17 @@ void onGoAwayReceived(int lastStreamId, long errorCode, ByteBuf debugData) {
+      synchronized (this) {
+        stream = streams.get(streamId);
+      }
+      System.out.println(getClass().getName() + ".onPriorityRead(" + new StreamPriority(streamDependency, weight, exclusive) + ") : " + stream);
```
remove System.out.println

vietj commented on src/main/java/io/vertx/core/http/impl/Http2ServerRequestImpl.java:
```diff
@@ -516,4 +520,25 @@ public HttpServerRequest customFrameHandler(Handler<HttpFrame> handler) {
+
+  @Override
+  void handlePriorityChange(StreamPriority streamPriority) {
+    System.out.println(getClass().getName() + ".handlePriorityChange(" + streamPriority + ")");
```
remove System.out.println

vietj commented on src/main/java/io/vertx/core/http/impl/Http2ServerRequestImpl.java:
```diff
@@ -516,4 +520,25 @@ public HttpServerRequest customFrameHandler(Handler<HttpFrame> handler) {
+
+  @Override
+  public HttpServerRequest streamPriorityHandler(Handler<StreamPriority> handler) {
+    streamPriorityHandler = handler;
```
wrap in `synchronized (conn)` block

vietj commented on src/main/java/io/vertx/core/http/impl/Http2ServerRequestImpl.java:
```diff
@@ -516,4 +520,25 @@ public HttpServerRequest customFrameHandler(Handler<HttpFrame> handler) {
+      if (handler != null) {
+        if(streamPriority != null && !streamPriority.equals(getStreamPriority())) {
+          setStreamPriority(streamPriority);
+          handler.handle(streamPriority);      
```
read the handler in the synchronized block but call handler outside of the synchronized block

vietj commented on src/main/java/io/vertx/core/http/impl/Http2ServerResponseImpl.java:
```diff
@@ -709,4 +712,15 @@ public HttpServerResponse push(HttpMethod method, String host, String path, Mult
+  public HttpServerResponse setStreamPriority(StreamPriority streamPriority) {
+    if(!streamPriority.equals(stream.getStreamPriority())) {
+      stream.setStreamPriority(streamPriority);
+      if(headWritten) {
```
I think it would be better to let the stream figure out or not whether it needs to send a priority frame internally. Indeed the HTTP/2 stream knows if the first header frame has been sent or not

mmopenet replied:
There is no variable in VertxHttp2Stream. I could not see the way to figure out if the head was written without adding state variable, so I though it will be simpler to reuse Http2ServerResponseImpl.headWritten.

vietj commented on src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java:
```diff
@@ -672,4 +679,25 @@ private void checkResponseHandler() {
+        this.streamPriority = streamPriority;
+        if(stream != null) {
+          stream.setStreamPriority(streamPriority);
+          stream.writePriorityFrame();
```
let the stream write the frame when `setStreamPriority` is called.

mmopenet replied:
Again the VertxHttp2Stream has not know is the headers frame was written.

vietj commented on src/main/java/io/vertx/core/http/impl/HttpClientResponseImpl.java:
```diff
@@ -189,6 +190,11 @@ public HttpClientResponse customFrameHandler(Handler<HttpFrame> handler) {
+  
+  @Override
+  public StreamPriority getStreamPriority() {
+    return stream.getStreamPriority();
```
wrap with `synchronized(conn)` block

vietj commented:
The PR is looking mostly good, there are a few changes to handle and the contributor agreement / commits to be signed

mmopenet commented:
Hi Julien,

I have signed ECA before making commits, but I missed the step to sign commits.
Should I resign the existing two commits. The only way I know to to that changes the history in the repository including the hashes of the commits.
I am not sure how that will affect this pull request and the review you have done so far?

Regards,
Michal

vietj commented:
redo the commits signed and make a new PR



> On 9 Nov 2018, at 07:54, Michal Michalowski <notifications@github.com> wrote:
> 
> Hi Julien,
> 
> I have signed ECA before making commits, but I missed the step to sign commits.
> Should I resign the existing two commits. The only way I know to to that changes the history in the repository including the hashes of the commits.
> I am not sure how that will affect this pull request and the review you have done so far?
> 
> Regards,
> Michal
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/eclipse-vertx/vert.x/pull/2693#issuecomment-437268033>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AANxiqgBcI4Y3mI90G8h4Lp-B3oJcepFks5utSbBgaJpZM4YLX_Y>.
>

mmopenet commented:
Hi Julien,

I have redone the commits signed off. The PR automatically picked them up so the IP validation now passed.
I have also addressed your comments to which I had no concerns and replied to others.

Regards,
Michal

vietj commented:
alright! I'll do another review / discussion

vietj commented:
@mmopenet to make API clearer on `HttpClientStream` we need to merge `setStreamPriority` and `writePriorityFrame` into a single method `updatePriority` that will always send a priority frame and have the `writeHead` method to take an extra `StreamPriority` that will send the priority along with the HTTP/2 header frame.

vietj commented on src/main/java/io/vertx/core/http/HttpClientResponse.java:
```diff
@@ -153,4 +153,18 @@
+   * Gets the priority of the associated HTTP/2 stream
+   * @return The priority of the associated HTTP/2 stream for HTTP/2 otherwise {@code null}  
+   */
+  StreamPriority getStreamPriority();
```
rename to `streamPriority()` as this is read only accessor

vietj commented on src/main/java/io/vertx/core/http/HttpServerRequest.java:
```diff
@@ -328,4 +329,19 @@ default HttpServerRequest bodyHandler(@Nullable Handler<Buffer> bodyHandler) {
+   * Gets the priority of the associated HTTP/2 stream
+   * @return The priority of the associated HTTP/2 stream for HTTP/2 otherwise {@code null}  
+   */
+  default StreamPriority getStreamPriority() {
```
rename to `streamPriority()` as this is a read-only accessor

vietj commented on src/main/java/io/vertx/core/http/StreamPriority.java:
```diff
@@ -0,0 +1,122 @@
+        this.exclusive = other.exclusive;
+    }
+
+    public short getWeight() {
```
javadoc

vietj commented on src/main/java/io/vertx/core/http/StreamPriority.java:
```diff
@@ -0,0 +1,122 @@
+        return weight;
+    }
+    
+    public int getStreamDependency() {
```
javadoc

vietj commented on src/main/java/io/vertx/core/http/StreamPriority.java:
```diff
@@ -0,0 +1,122 @@
+        return streamDependency;
+    }
+    
+    public boolean isExclusive() {
```
javadoc

vietj commented on src/main/java/io/vertx/core/http/StreamPriority.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
I think equals method can be written more concisely

mmopenet replied:
The method was generated automatically by Eclipse. I'll change the style so follows the convention in project.

vietj commented on src/main/java/io/vertx/core/http/impl/Http2ClientConnection.java:
```diff
@@ -132,6 +133,14 @@ private void recycle() {
   }
 
+  @Override
+  public synchronized void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers, int streamDependency, short weight, boolean exclusive, int padding, boolean endOfStream) throws Http2Exception {
```
rather change `stream#handleHeaders` to be `handleHeaders(Http2Headers,StreamPriority,boolean)` and handle this the same way than `onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers, int padding, boolean endOfStream)` do

mmopenet commented:
Yes, the spec is not clear on that. In section 5.3 it says:

> _A client **can** assign a priority for a new stream by including prioritization information in the HEADERS frame_ 

Then it also says:

> _At any other time, the PRIORITY frame (Section 6.3) can be used to change the priority of a stream._

Then:

> _The purpose of prioritization is to allow an **endpoint** to express how it would prefer its peer to allocate resources when managing concurrent streams_

In addition the section 6.3:

> _The PRIORITY frame (type=0x2) specifies the **sender-advised** priority of a stream (Section 5.3). It can be sent in any stream state, including idle or closed streams._

So it first allows the **client** to set the priority for the newly created stream, and then it allows an **endpoint** to change the priority using PRIORITY frame. It also says in section 6.2;
> _Prioritization information in a HEADERS frame is logically equivalent to a separate PRIORITY frame_
My reading of that is that the server as an endpoint can change the priority of the stream using either HEADERS or PRIORITY frame.

vietj commented:
@mmopenet I think it allows the client to change the priority after it has set it. The use case is that browsers are very dynamic and they discover resource priorities incrementally (when they download new resources or execute more javascript), so it can change the priority after it has initiated a stream.

So I would like that we  remove the features on the server because it's confusing and later we can still add it if there is an actual use case for it.

mmopenet commented:
@vietj There is already usecase in 5G Core network related to overload management. The messages are prioritized depending on the importance, e.g. multimedia priority sessions, emergency calls, etc.
Have a look at: http://www.3gpp.org/ftp//Specs/archive/29_series/29.500/29500-f10.zip
There is explicit mention of changes priorities by server as well.
I agree that the original usecase for that was usage in browsers, but it has more usecases now.

vietj commented:
so that means the server would override the priority stream the client has set ?

as well as the dependency ?

where is it in the document ?

> On 16 Nov 2018, at 11:41, Michal Michalowski <notifications@github.com> wrote:
> 
> @vietj <https://github.com/vietj> There is already usecase in 5G Core network related to overload management. The messages are prioritized depending on the importance, e.g. multimedia priority sessions, emergency calls, etc.
> Have a look at: http://www.3gpp.org/ftp//Specs/archive/29_series/29.500/29500-f10.zip <http://www.3gpp.org/ftp//Specs/archive/29_series/29.500/29500-f10.zip>
> There is explicit mention of changes priorities by server as well.
> I agree that the original usecase for that was usage in browsers, but it has more usecases now.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/eclipse-vertx/vert.x/pull/2693#issuecomment-439354226>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AANxijAL1y3EDOv03No_BwfbIMWjZmDQks5uvpZWgaJpZM4YLX_Y>.
>

mmopenet commented:
> so that means the server would override the priority stream the client has set ? as well as the dependency ? where is it in the document ?
> […](#)
> On 16 Nov 2018, at 11:41, Michal Michalowski ***@***.***> wrote: @vietj <https://github.com/vietj> There is already usecase in 5G Core network related to overload management. The messages are prioritized depending on the importance, e.g. multimedia priority sessions, emergency calls, etc. Have a look at: http://www.3gpp.org/ftp//Specs/archive/29_series/29.500/29500-f10.zip <http://www.3gpp.org/ftp//Specs/archive/29_series/29.500/29500-f10.zip> There is explicit mention of changes priorities by server as well. I agree that the original usecase for that was usage in browsers, but it has more usecases now. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#2693 (comment)](https://github.com/eclipse-vertx/vert.x/pull/2693#issuecomment-439354226)>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AANxijAL1y3EDOv03No_BwfbIMWjZmDQks5uvpZWgaJpZM4YLX_Y>.

Yes the server can change the priority of the stream if the response message shoudl have lover priority that the request.

In the 3GPP TS 29.500 6.8.3

> If the server wants to change the priority of the response, it shall send a PRIORITY frame after the stream state became "half-closed (remote)".

The stream dependency is always set to 0 in this usecase. The purpose is to use the priority weight to prioritized all messages at a connection level.

vietj commented:
ok makes sense, you convinced me :-)

mmopenet commented:
:)

vietj commented:
one small change : on `HttpClientStream` rename `streamPriority()` and `updateStreamPriority` to just `priority` and `updatePriority` since this is an operation on a stream so implicitely being already a stream

mmopenet commented:
OK. 
What about VertxHttp2Stream.setStreamPriority, VertxHttp2Stream.streamPriority?
Woudl you like to change them to VertxHttp2Stream.setPriority and VertxHttp2Stream.priority
or Both to VertxHttp2Stream.priority.

mmopenet commented:
@vietj But the methods in VertxHttp2Stream are also inherited by Http2ServerRequestImpl. So to avoid heaving Http2ServerRequestImpl with priority and streamPriority I would need to modify HttpServerRequest, to have priority method.

vietj commented:
you can have these methods as delegate, i.e have setStreamPriority on Http2ServerRequestImpl call this.setPriority

> On 16 Nov 2018, at 16:18, Michal Michalowski <notifications@github.com> wrote:
> 
> @vietj <https://github.com/vietj> But the methods in VertxHttp2Stream are also inherited by Http2ServerRequestImpl. So to avoid heaving Http2ServerRequestImpl with priority and streamPriority I would need to modify HttpServerRequest, to have priority method.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/eclipse-vertx/vert.x/pull/2693#issuecomment-439425346>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AANxil8Eut8S4w54xXAiEGAeGS1YCmqcks5uvtctgaJpZM4YLX_Y>.
>

vietj commented:
yes if that's possible, the most important point was rather for the user facing API that can't be changed later.

mmopenet commented:
@vietj OK, I have updated the HttpClientStream and VertxHttp2Stream. The Response/Request interfaces still have set/getStreamPriority.

vietj commented:
I will try to merge it today before 3.6.0.CR2 release
